lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +104,149 @@ pub fn handle_completion(
+                server.reply(Response::new_ok(id, Value::Null));
+                return Ok(());
+            }
+            let name_id = name_id.unwrap();
+
+            linearization
+                .get_in_scope(&item)
+                .iter()
+                .filter_map(|i| {
+                    let (ty, _) = linearization.resolve_item_type_meta(i);
+                    match i.kind {
+                        // Get record fields from static type info
+                        TermKind::Declaration(..)
+                            if name_id == i.id && matches!(ty, Types(AbsType::Record(..))) =>
+                        {
+                            match &ty {
+                                Types(AbsType::Record(row)) => {
+                                    Some((extract_ident(row), i.ty.clone()))
+                                }
+                                _ => unreachable!(),
+                            }
+                        }
+                        // Get record fields from contract metadata
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id
+                                && find_record_contract(&linearization.linearization, body_id)
+                                    .is_some() =>
+                        {
+                            // unwrapping is safe here because of the pattern condition.
+                            let contract =
+                                find_record_contract(&linearization.linearization, body_id)
+                                    .unwrap();
+                            let fields = match &contract.types {
+                                Types(AbsType::Record(row)) => extract_ident(row),
+                                _ => Vec::with_capacity(0),
+                            };
+                            Some((fields, i.ty.clone()))
+                        }
+
+                        // Get record fields from lexical scoping
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id =>
+                        {
+                            find_record_fields(&linearization.linearization, body_id)
+                                .map(|idents| Some((idents, i.ty.clone())))
+                                .unwrap_or(None)
+                        }
+
+                        _ => None,
+                    }
+                })
+                .collect()
+        }
+
+        // variable name completion
+        Some(..) | None => linearization
+            .get_in_scope(&item)
+            .iter()
+            .filter_map(|i| match i.kind {
+                TermKind::Declaration(ref ident, _, _) => Some((vec![ident.clone()], i.ty.clone())),
+                _ => None,
+            })
+            .collect::<Vec<_>>(),
+    };
+
+    // O(n^2)
+    fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
```

Comment:
What are the benefits of declaring this function inside `handle_completion` function, rather than on its own?

Doesn't it create a new function each time `handle_completion` is called?

Reply:
IIRC, such functions are actually lifted to the top-level (because the compiled code has no notion of "local" functions). So in term of code generation, this should be strictly equivalent. Usually you use such local functions that because for specialized helpers, just as a way of better organizing your function implementation. Your helper may not make sense in any other context, or you don't expect it to be reused somewhere else. In this case, a local function don't pollute the top-level module API.

Reply:
My comment was more about code organization and clarity than performance or compilation :sweat_smile: 

Having this function **in the middle** of the algorithm feels weird and disrupts the reader in my opinion. Since the function name is rather clear, there isn't a need to understand that function implementation to understand the algorithm here. So by moving it out of the algorithm (or moving it at the very beginning before the start of the algorithm) the algorithm would be clearer for readers

Reply:
But you asked 

> Doesn't it create a new function each time handle_completion is called?

which I interpreted as does it allocate a new closure each time the function is called? What did you originally mean by that, then?

> Having this function in the middle of the algorithm feels weird and disrupts the reader in my opinion. Since the function name is rather clear, there isn't a need to understand that function implementation to understand the algorithm here. So by moving it out of the algorithm (or moving it at the very beginning before the start of the algorithm) the algorithm would be clearer for readers

Ah, then yes, I totally agree. Subfunctions should be declared only at the beginning of a function, IMHO.